release-20.2: colflow: cancel flow on ungraceful stream shutdown in outbox

yuzefovich opened the pull request:
Backport 1/1 commits from #63772.
Backport 1/1 commits from #64386.

/cc @cockroachdb/release

---

**colflow: cancel flow on ungraceful stream shutdown in outbox**

Previously, when an outbox `Recv`ed a non-`io.EOF` error from the gRPC
stream, we would simply log it. Such an error indicates an ungraceful
shutdown of the stream, and, instead, it should lead to the cancellation
of the whole stream. Previously, the main goroutine of the outbox would
keep on running and it would exit only the next time it attempted to
`Send` something on the stream. This is now fixed, and the flow ctx will
be canceled when a stream is shutdown ungracefully. This is what we do
in the row-based flows, and this commit adds this behavior the
vectorized outbox.

Fixes: https://github.com/cockroachlabs/support/issues/924.

Release note (bug fix): Previously, the remote flows of execution in the
vectorized engine could take quite a long time to shut down whenever a
node participating in the plan dies.

**colrpc: mark flow ctx cancellation in inbox as ungraceful**

We have recently fixed the shutdown of the gRPC streams used by DistSQL.
However, during the backport of that commit (which wasn't clean) I made
a suboptimal decision of treating the flow context cancellation during
the inbox setup as "graceful" termination of the stream - I believe it
should be treated as ungraceful (and this what we'll have on master).
This commit makes it so.

The difference between two is that if the termination of the stream is
ungraceful, the outbox (the client of the FlowStream RPC) will cancels
the flow context of its host whereas if the termination is graceful, the
outbox will only shutdown its own tree of components (by canceling its
own context). These things are equivalent from the correctness point of
view, but the former is more optimal way of shutting things down.

Release note: None

yuzefovich committed:
colflow: cancel flow on ungraceful stream shutdown in outbox

Previously, when an outbox `Recv`ed a non-`io.EOF` error from the gRPC
stream, we would simply log it. Such an error indicates an ungraceful
shutdown of the stream, and, instead, it should lead to the cancellation
of the whole stream. Previously, the main goroutine of the outbox would
keep on running and it would exit only the next time it attempted to
`Send` something on the stream. This is now fixed, and the flow ctx will
be canceled when a stream is shutdown ungracefully. This is what we do
in the row-based flows, and this commit adds this behavior the
vectorized outbox.

Release note (bug fix): Previously, the remote flows of execution in the
vectorized engine could take quite a long time to shut down whenever a
node participating in the plan dies.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64220)
<!-- Reviewable:end -->

yuzefovich commented:
I cherry-picked the commit from the backport to 21.1 (#64219).

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64220#-:-MZLBgY0364lTp_svEbj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colrpc: mark flow ctx cancellation in inbox as ungraceful

We have recently fixed the shutdown of the gRPC streams used by DistSQL.
However, during the backport of that commit (which wasn't clean) I made
a suboptimal decision of treating the flow context cancellation during
the inbox setup as "graceful" termination of the stream - I believe it
should be treated as ungraceful (and this what we'll have on master).
This commit makes it so.

The difference between two is that if the termination of the stream is
ungraceful, the outbox (the client of the FlowStream RPC) will cancels
the flow context of its host whereas if the termination is graceful, the
outbox will only shutdown its own tree of components (by canceling its
own context). These things are equivalent from the correctness point of
view, but the former is more optimal way of shutting things down.

Release note: None

yuzefovich commented:
I included the backport of #64386 into this PR as well to be merged together, PTAL.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64220#-:-MZmtodA5tKNpcAUhyb0:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->
